File: app/src/main/java/com/nfcalarmclock/NacSnoozeDurationPreference.java
package com.nfcalarmclock;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import androidx.fragment.app.FragmentManager;
import androidx.preference.Preference;

/**
 * Preference that displays how long to snooze for.
 */
public class NacSnoozeDurationPreference
	extends Preference
	implements NacScrollablePickerDialogFragment.OnScrollablePickerOptionSelectedListener
{

	/**
	 * Preference value.
	 */
	private int mValue;

	/**
	 */
	public NacSnoozeDurationPreference(Context context)
	{
		this(context, null);
	}

	/**
	 */
	public NacSnoozeDurationPreference(Context context, AttributeSet attrs)
	{
		this(context, attrs, 0);
	}

	/**
	 */
	public NacSnoozeDurationPreference(Context context, AttributeSet attrs,
		int style)
	{
		super(context, attrs, style);
		setLayoutResource(R.layout.nac_preference);
	}

	/**
	 * @return The summary text.
	 */
	@Override
	public CharSequence getSummary()
	{
		NacSharedConstants cons = new NacSharedConstants(getContext());
		int value = this.getValue();

		return NacSharedPreferences.getSnoozeDurationSummary(cons, value);
	}

	/**
	 * Get the preference value.
	 *
	 * @return The preference value.
	 */
	public int getValue()
	{
		return this.mValue;
	}

	/**
	 * @return The default value.
	 */
	@Override
	protected Object onGetDefaultValue(TypedArray a, int index)
	{
		NacSharedDefaults defs = new NacSharedDefaults(getContext());
		return a.getInteger(index, defs.getSnoozeDurationIndex());
	}

	/**
	 * Save the selected value from the scrollable picker.
	 */
	@Override
	public void onScrollablePickerOptionSelected(int index)
	{
		this.mValue = index;

		persistInt(index);
		notifyChanged();
	}

	/**
	 * Set the initial preference value.
	 */
	@Override
	protected void onSetInitialValue(Object defaultValue)
	{
		if (defaultValue == null)
		{
			this.mValue = getPersistedInt(this.mValue);
		}
		else
		{
			this.mValue = (Integer) defaultValue;
			persistInt(this.mValue);
		}
	}

	/**
	 * Show the auto dismiss dialog.
	 */
	public void showDialog(FragmentManager manager)
	{
		NacSnoozeDurationDialog dialog = new NacSnoozeDurationDialog();
		int value = this.getValue();

		dialog.setDefaultScrollablePickerIndex(value);
		dialog.setOnScrollablePickerOptionSelectedListener(this);
		dialog.show(manager, NacSnoozeDurationDialog.TAG);
	}

}

File: app/src/main/java/com/nfcalarmclock/snoozeduration/NacSnoozeDurationPreference.kt
package com.nfcalarmclock.snoozeduration

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import androidx.fragment.app.FragmentManager
import androidx.preference.Preference
import com.nfcalarmclock.R
import com.nfcalarmclock.shared.NacSharedConstants
import com.nfcalarmclock.shared.NacSharedPreferences
import com.nfcalarmclock.view.dialog.NacScrollablePickerDialogFragment.OnScrollablePickerOptionSelectedListener

/**
 * Preference that displays how long to snooze for.
 */
class NacSnoozeDurationPreference @JvmOverloads constructor(

	/**
	 * Context.
	 */
	context: Context,

	/**
	 * Attribute set.
	 */
	attrs: AttributeSet? = null,

	/**
	 * Default style.
	 */
	style: Int = 0

	// Constructor
) : Preference(context, attrs, style),

	// Interface
	OnScrollablePickerOptionSelectedListener
{

	/**
	 * Preference value.
	 */
	private var snoozeDurationIndex = 0

	/**
	 * Constructor.
	 */
	init
	{
		layoutResource = R.layout.nac_preference
	}

	/**
	 * Get the summary text.
	 *
	 * @return The summary text.
	 */
	override fun getSummary(): CharSequence
	{
		val cons = NacSharedConstants(context)

		return NacSharedPreferences.getSnoozeDurationSummary(cons, snoozeDurationIndex)
	}

	/**
	 * Get the default value.
	 *
	 * @return The default value.
	 */
	override fun onGetDefaultValue(a: TypedArray, index: Int): Any
	{
		val defaultValue = context.resources.getInteger(R.integer.default_snooze_duration_index)

		return a.getInteger(index, defaultValue)
	}

	/**
	 * Save the selected value from the scrollable picker.
	 */
	override fun onScrollablePickerOptionSelected(index: Int)
	{
		// Set the snooze duration index
		snoozeDurationIndex = index

		// Persist the index
		persistInt(index)

		// Notify of a change
		notifyChanged()
	}

	/**
	 * Set the initial preference value.
	 */
	override fun onSetInitialValue(defaultValue: Any?)
	{
		// Check if the default value is null
		if (defaultValue == null)
		{
			snoozeDurationIndex = getPersistedInt(snoozeDurationIndex)
		}
		// Convert the default value
		else
		{
			snoozeDurationIndex = defaultValue as Int

			persistInt(snoozeDurationIndex)
		}
	}

	/**
	 * Show the auto dismiss dialog.
	 */
	fun showDialog(manager: FragmentManager)
	{
		// Create the dialog
		val dialog = NacSnoozeDurationDialog()

		// Setup the dialog
		dialog.defaultScrollablePickerIndex = snoozeDurationIndex
		dialog.onScrollablePickerOptionSelectedListener = this

		// Show the dialog
		dialog.show(manager, NacSnoozeDurationDialog.TAG)
	}

}